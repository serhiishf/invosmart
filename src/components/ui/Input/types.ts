import { FunctionComponent } from 'react';

export interface InputProps {
  id?: string;
  type?: string;
  name?: string;
  label?: string;
  value?: string;
  placeholder?: string;
  required?: boolean;
  disabled?: boolean;
  readonly?: boolean;
  isValid?: boolean;
  isError?: boolean;
  errorMessage?: string;
  onChange?: (value: string) => void;
  Icon?: FunctionComponent;
}
