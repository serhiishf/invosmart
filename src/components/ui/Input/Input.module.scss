.input {
  width: 100%;
  height: 48px;
  @include for-size-up(desktop) {
    height: 40px;
  }
}

.prefixContainer {
  padding-left: 25px;
  padding-right: 8px;
  display: flex;
  justify-content: center;
  align-items: center;
}

.staticPrefixIcon {
  display: flex;
  justify-content: center;
  align-items: center;
  color: $inputStaticIconColor;
}

.inputBaseProps {
  padding-left: 0px;
}

.suffixContainer {
  display: flex;
  justify-content: center;
  align-items: center;
  padding-left: 8px;
  padding-right: 4px;
  gap: 4px;
}

.attentionContainer {
  position: relative;
  display: flex;
  justify-content: center;
  align-items: center;
  padding: 8px;
}

.attentionIcon {
  color: $inputAttentionIconColor;
}

@media (pointer: fine) {
  .attentionContainer:hover .tooltipWrap {
    visibility: visible;
    opacity: 1;
    transition: all $inputAnimation;
  }
}

.attentionContainer:active .tooltipWrap {
  visibility: visible;
  opacity: 1;
  transition: all $inputAnimation;
}

.tooltipWrap {
  position: absolute;
  visibility: hidden;
  opacity: 0;
  max-width: 260px;
  width: max-content;
  height: auto;
  top: 35px;
  right: -5px;
  z-index: $inputTooltipZIndex;
  transition: all $inputAnimation;
  @include for-size-up(desktop) {
    top: 35px;
  }
}

.passwordButtonContainer {
  display: flex;
  justify-content: center;
  align-items: center;
}

.passwordButtonWrap {
  color: $inputPasswordButtonColorDefault;
  width: 42px;
  transition: all $inputAnimation;
  &:hover {
    color: $inputPasswordButtonColorHover;
    transition: all $inputAnimation;
  }
  &:active {
    color: $inputPasswordButtonColorActive;
  }
  @include for-size-up(desktop) {
    width: 34px;
  }
}

.passwordButtonIcon {
  @include for-size-up(desktop) {
    scale: 80%;
  }
}
