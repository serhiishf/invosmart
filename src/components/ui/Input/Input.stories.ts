import type { Meta, StoryObj } from '@storybook/react';
import { fn } from '@storybook/test';
import Input from './Input';


const meta = {
  title: 'components/UI/Input',
  component: Input,
  tags: ['autodocs'],
  args: {
    onChange: fn(),
  },
} satisfies Meta<typeof Input>;

export default meta;

type Story = StoryObj<typeof meta>;
export const Default: Story = {};

export const WithLabelAndPlaceholder: Story = {
  args: {
    label: 'Label',
    placeholder: 'Placeholder',
  },
};

export const Password: Story = {
  args: {
    label: 'Password',
    type: 'password',
    placeholder: 'Enter password',
  },
};

export const Error: Story = {
  args: {
    label: 'Error state',
    placeholder: 'Error state',
    isError: true,
  },
};

export const ReadOnlyMode: Story = {
  args: {
    label: 'Readonly mode',
    placeholder: 'readOnly',
    readOnly: true,
  },
};
