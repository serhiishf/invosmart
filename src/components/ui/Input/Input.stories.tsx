import type { Meta, StoryObj } from '@storybook/react';
import { fn } from '@storybook/test';
import Input from './Input';
import { iconSelectConfig } from 'config/storybookIconConfig';
import IconSearch from 'assets/icons/search.svg?react';

const meta = {
  title: 'components/UI/Input',
  component: Input,
  tags: ['autodocs'],
  argTypes: {
    PrefixIcon: {
      ...iconSelectConfig,
    },
  },
  args: {
    onChange: fn(),
  },
} satisfies Meta<typeof Input>;

export default meta;

type Story = StoryObj<typeof meta>;
export const Default: Story = {};

export const Password: Story = {
  args: {
    label: 'Password',
    type: 'password',
    placeholder: 'Enter password placeholder',
  },
};

export const Error: Story = {
  args: {
    label: 'Error state',
    placeholder: 'Placeholder',
    isError: true,
  },
};

export const ReadonlyMode: Story = {
  args: {
    label: 'Readonly mode',
    value: 'Readonly mode value',
    placeholder: 'Placeholder',
    readOnly: true,
  },
};

export const Disabled: Story = {
  args: {
    label: 'Disabled',
    value: 'Disabled value',
    placeholder: 'Placeholder',
    disabled: true,
  },
};

export const WithPrefixIcon: Story = {
  args: {
    label: 'With Prefix Icon',
    placeholder: 'Placeholder',
    PrefixIcon: IconSearch,
  },
};

export const ThemeBackground: Story = {
  decorators: [
    (Story) => (
      <div style={{ background: '#f5f6fa', padding: '50px' }}>
        <Story />
      </div>
    ),
  ],
  args: {
    label: 'Theme bg',
    placeholder: 'Placeholder',
  },
};
