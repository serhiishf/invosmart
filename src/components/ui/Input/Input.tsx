import { useState } from 'react';
import { FieldWrapper, InputBase, Tooltip } from '../';
import { InputProps } from './types';
import styles from './Input.module.scss';
import AttentionIcon from '../../../assets/icons/attention.svg?react';
import classNames from 'classnames';

function Input(props: InputProps) {
  const {
    id,
    type = 'text',
    name,
    label,
    value,
    placeholder,
    required = false,
    disabled = false,
    readonly = false,
    isValid = true,
    isError,
    errorMessage,
    onChange,
    Icon,
  } = props;

  const [isFocused, setIsFocused] = useState(false);

  function onFocusHandler() {
    setIsFocused(true);
  }

  function onBlurHandler() {
    setIsFocused(false);
  }

  return (
    <div className={styles.input}>
      <FieldWrapper
        label={label}
        focus={isFocused}
        error={!isValid || isError}
        readonly={readonly}
        disabled={disabled}
      >
        <>
          {Icon && (
            <div className={styles.staticIconLeft}>
              <Icon />
            </div>
          )}
        </>
        <InputBase
          id={id}
          type={type}
          name={name}
          required={required}
          disabled={disabled}
          readonly={readonly}
          value={value}
          placeholder={placeholder}
          onFocus={onFocusHandler}
          onBlur={onBlurHandler}
          onChange={onChange}
          className={classNames(!!Icon && styles.inputBaseProps)}
        />
        <>
          {isError && (
            <div className={styles.attentionContainer}>
              <AttentionIcon className={styles.attentionContainer__icon} />
            </div>
          )}
        </>
        <>
          {errorMessage && (
            <div className={styles.tooltipContainer}>
              <Tooltip tooltipMessage={errorMessage} arrow="topRight" isError={isError} />
            </div>
          )}
        </>
      </FieldWrapper>
    </div>
  );
}

export default Input;
