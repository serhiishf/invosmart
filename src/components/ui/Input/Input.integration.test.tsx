import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Input from './Input';

describe('Input', () => {
  describe('default Initialization', () => {
    it('should match the snapshot with default props', () => {
      const { asFragment } = render(<Input />);
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('initialization with props', () => {
    it('default', () => {});
  });
});
