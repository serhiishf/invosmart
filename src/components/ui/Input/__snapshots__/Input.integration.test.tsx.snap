// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Input > default Initialization > should match the snapshot with default props 1`] = `
<DocumentFragment>
  <div
    class="_input_ab7e9c"
    data-testid="input-component"
  >
    <div
      class="_fieldWrapper_3d96c9"
    >
      <div
        class="_fieldWrapper__border_3d96c9"
      >
        <input
          class="_inputBase_6ca9ba _inputBase--paddingLeft-default_6ca9ba _inputBase--paddingRight-default_6ca9ba"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
